File: typescript/bc-lifehash/src/cell-grid.ts
import { BitAggregator, BitEnumerator } from './bit-enumerator.js';
import { ChangeGrid } from './change-grid.js';
import { Grid } from './grid.js';

export class CellGrid {
    readonly grid: Grid<boolean>;

    constructor(width: number, height: number) {
        this.grid = new Grid(width, height, false);
    }

    private static isAliveInNextGeneration(
        currentAlive: boolean,
        neighborsCount: number,
    ): boolean {
        if (currentAlive) {
            return neighborsCount === 2 || neighborsCount === 3;
        }
        return neighborsCount === 3;
    }

    private countNeighbors(px: number, py: number): number {
        let total = 0;
        this.grid.forNeighborhood(px, py, (ox, oy, nx, ny) => {
            if (ox === 0 && oy === 0) return;
            if (this.grid.getValue(nx, ny)) {
                total += 1;
            }
        });
        return total;
    }

    data(): Uint8Array {
        const a = new BitAggregator();
        this.grid.forAll((x, y) => {
            a.append(this.grid.getValue(x, y));
        });
        return a.data();
    }

    setData(data: Uint8Array): void {
        const e = new BitEnumerator(data);
        let i = 0;
        e.forAll((b) => {
            this.grid.storage[i] = b;
            i += 1;
        });
    }

    nextGeneration(
        currentChangeGrid: ChangeGrid,
        nextCellGrid: CellGrid,
        nextChangeGrid: ChangeGrid,
    ): void {
        nextCellGrid.grid.setAll(false);
        nextChangeGrid.grid.setAll(false);
        const width = this.grid.width;
        const height = this.grid.height;
        for (let y = 0; y < height; y++) {
            for (let x = 0; x < width; x++) {
                const currentAlive = this.grid.getValue(x, y);
                if (currentChangeGrid.grid.getValue(x, y)) {
                    const neighborsCount = this.countNeighbors(x, y);
                    const nextAlive = CellGrid.isAliveInNextGeneration(
                        currentAlive,
                        neighborsCount,
                    );
                    if (nextAlive) {
                        nextCellGrid.grid.setValue(true, x, y);
                    }
                    if (currentAlive !== nextAlive) {
                        nextChangeGrid.setChanged(x, y);
                    }
                } else {
                    nextCellGrid.grid.setValue(currentAlive, x, y);
                }
            }
        }
    }
}

File: csharp/BCLifeHash/BCLifeHash/CellGrid.cs
using System.Diagnostics;

namespace BlockchainCommons.BCLifeHash;

internal sealed class CellGrid
{
    public Grid<bool> Grid { get; }

    public CellGrid(int width, int height)
    {
        Grid = new Grid<bool>(width, height);
    }

    private static bool IsAliveInNextGeneration(bool currentAlive, int neighborsCount)
    {
        if (currentAlive)
            return neighborsCount == 2 || neighborsCount == 3;
        return neighborsCount == 3;
    }

    private int CountNeighbors(int px, int py)
    {
        var total = 0;
        Grid.ForNeighborhood(px, py, (ox, oy, nx, ny) =>
        {
            if (ox == 0 && oy == 0)
                return;
            if (Grid.GetValue(nx, ny))
                total++;
        });
        return total;
    }

    public byte[] GetData()
    {
        var a = new BitAggregator();
        Grid.ForAll((x, y) =>
        {
            a.Append(Grid.GetValue(x, y));
        });
        return a.GetData();
    }

    public void SetData(byte[] data)
    {
        Debug.Assert(Grid.Width * Grid.Height == data.Length * 8);
        var e = new BitEnumerator(data);
        var i = 0;
        e.ForAll(b =>
        {
            Grid.Storage[i] = b;
            i++;
        });
        Debug.Assert(i == Grid.Storage.Length);
    }

    public void NextGeneration(
        ChangeGrid currentChangeGrid,
        CellGrid nextCellGrid,
        ChangeGrid nextChangeGrid)
    {
        nextCellGrid.Grid.SetAll(false);
        nextChangeGrid.Grid.SetAll(false);
        var width = Grid.Width;
        var height = Grid.Height;
        for (var y = 0; y < height; y++)
        {
            for (var x = 0; x < width; x++)
            {
                var currentAlive = Grid.GetValue(x, y);
                if (currentChangeGrid.Grid.GetValue(x, y))
                {
                    var neighborsCount = CountNeighbors(x, y);
                    var nextAlive = IsAliveInNextGeneration(currentAlive, neighborsCount);
                    if (nextAlive)
                        nextCellGrid.Grid.SetValue(true, x, y);
                    if (currentAlive != nextAlive)
                        nextChangeGrid.SetChanged(x, y);
                }
                else
                {
                    nextCellGrid.Grid.SetValue(currentAlive, x, y);
                }
            }
        }
    }
}
